typescript/cli/src/rebalancer/monitor/Monitor.ts

```diff
@@ -26,51 +24,74 @@ export class Monitor implements IMonitor {
     private readonly checkFrequency: number,
   ) {}
 
-  subscribe(fn: (data: MonitorEvent) => void) {
-    this.emitter.on(this.MONITOR_EVENT, fn);
+  on(
+    eventName: 'collateralbalances' | 'start' | 'error',
+    fn: (...args: any[]) => void,
+  ) {
+    this.emitter.on(eventName, fn);
+    return this;
   }
 
   async start() {
-    if (this.interval) {
+    if (this.isMonitorRunning) {
       // Cannot start the same monitor multiple times
-      throw new Error('Monitor already running');
+      this.emitter.emit('error', new Error('Monitor already running'));
     }
 
-    // Build the WarpCore from the registry
-    const metadata = await this.registry.getMetadata();
-    const addresses = await this.registry.getAddresses();
-    const mailboxes = objMap(addresses, (_, { mailbox }) => ({ mailbox }));
-    const provider = new MultiProtocolProvider(objMerge(metadata, mailboxes));
-    const warpCoreConfig = await this.registry.getWarpRoute(this.warpRouteId);
-    const warpCore = WarpCore.FromConfig(provider, warpCoreConfig);
-
-    // Start the interval used to poll collateral balances
-    this.interval = setInterval(async () => {
-      const event: MonitorEvent = {
-        balances: [],
-      };
-
-      for (const token of warpCore.tokens) {
-        // Ignore non-collateralized tokens given that we only care about collateral balances
-        if (!token.isCollateralized()) {
-          continue;
-        }
+    try {
+      this.isMonitorRunning = true;
+
+      // Build the WarpCore from the registry
+      const metadata = await this.registry.getMetadata();
+      const addresses = await this.registry.getAddresses();
+
+      // The Sealevel warp adapters require the Mailbox address, so we
+      // get mailboxes for all chains and merge them with the chain metadata.
+      const mailboxes = objMap(addresses, (_, { mailbox }) => ({ mailbox }));
+      const provider = new MultiProtocolProvider(objMerge(metadata, mailboxes));
+      const warpCoreConfig = await this.registry.getWarpRoute(this.warpRouteId);
+      const warpCore = WarpCore.FromConfig(provider, warpCoreConfig);
+
+      // this can be considered the starting point of the monitor
+      this.emitter.emit('start');
+
+      while (this.isMonitorRunning) {
+        const event: MonitorEvent = {
```

Comment:
The content in here should be wrapped inside a try to separate start failures from polling failures.

start failures should exit the process while poll errors should just be logged

Reply:
and this is completely related to my previous comment; ok, will give it a go here if it's cheap